src/DurableTask.ApplicationInsights/DurableTelemetryModule.cs

```
@@ -0,0 +1,138 @@
+    /// </summary>
+    public sealed class DurableTelemetryModule : ITelemetryModule, IAsyncDisposable
+    {
+        private TelemetryClient telemetryClient;
+        private ActivityListener listener;
+
+        /// <inheritdoc/>
+        public void Initialize(TelemetryConfiguration configuration)
+        {
+            this.telemetryClient = new TelemetryClient(configuration);
+
+            // ActivitySamplingResult.AllData means that the ActivityListener is going to collect all of the data
+            // for any Activity that's sent to the "DurableTask" source. It isn't going to exclude any data.
+            this.listener = new ActivityListener
+            {
+                ShouldListenTo = source => source.Name.StartsWith("DurableTask"),
+                ActivityStopped = OnEndActivity,
+                Sample = (ref ActivityCreationOptions<ActivityContext> _) => ActivitySamplingResult.AllData,
+                SampleUsingParentId = (ref ActivityCreationOptions<string> _) => ActivitySamplingResult.AllData
+            };
+
+            ActivitySource.AddActivityListener(this.listener);
+        }
+
+        void OnEndActivity(Activity activity)
+        {
+            if (!activity.IsAllDataRequested)
+            {
+                return;
+            }
+
+            OperationTelemetry telemetry = CreateTelemetry(activity);
+            this.telemetryClient.Track(telemetry);
+        }
+
+        static OperationTelemetry CreateTelemetry(Activity activity)
+        {
+            OperationTelemetry telemetry;
+            ActivityStatusCode status = activity.GetStatus(out string description);
+            switch (activity.Kind)
+            {
+                case ActivityKind.Consumer or ActivityKind.Server:
+                    RequestTelemetry request = CreateTelemetryCore<RequestTelemetry>(activity);
+                    request.Success = status != ActivityStatusCode.Error;
```

For an added layer of resiliency, can you add a check for `request.StatusCode ??= (request.Success ? "success" : "fail");`. This is to guard against `durabletask.task.status` not being preset and we never set `StatusCode` (which if not set app insights silently fails to send this up).

Also instead of the raw strings `success` and `fail` - set it to the equivalent of `durabletask.task.status` in success/fail cases.

---

I ended up adding the following check. I replaced `??=` with a null check because `request.ResponseCode` was never `null` and evaluated to an empty string - `""`. The `bool` cast is there for `request.Success` because it's nullable.

```
if (string.IsNullOrEmpty(request.ResponseCode))
{
    request.ResponseCode = (bool)request.Success ? "200" : "500";
}
```